packages/stylelint-plugin/src/rules/no-invalid-functions.ts

```diff
@@ -0,0 +1,360 @@
+    if (!variable) return original
+    const replacementVariable =
+      getAndReportOnReplacementRgbParamsVariable(variable, decl, options) ||
+      first
+
+    const fixedSecondParameter = parsePercentageOrDecimal(second)
+
+    if (fixedSecondParameter === "NaN") return original
+
+    return `rgba(${replacementVariable}, ${1 - fixedSecondParameter})`
+  }
+
+  const compileSassValueInPlace = (
+    functionName: string,
+    ...params: string[]
+  ) => {
+    const sourceValue = `${functionName}(${params.join(", ")})`
+
+    const compileResult = compileSassValue(decl.root(), decl, sourceValue)
+    if (compileResult.error || !compileResult.compiledValue) {
+      options.reporter({
+        autofixAvailable: false,
+        message: unableToCompileFunctionMessage(
+          sourceValue,
+          (compileResult.details instanceof Error
+            ? compileResult.details.message
+            : compileResult.error || "unknown reason"
+          ).replace(/\s+/g, " ")
+        ),
+        node: decl,
+      })
+      return sourceValue
+    } else {
+      if (options.fix && !isVariable(decl)) {
+        return compileResult.compiledValue
+      } else {
+        options.reporter({
+          autofixAvailable: !isVariable(decl),
+          message: unsupportedFunctionWithFixMessage,
+          node: decl,
+        })
+        return sourceValue
+      }
+    }
+  }
+
+  const newValue = transformDecl(decl.value, {
+    rgba: fixRgbAndRgba,
+    rgb: fixRgbAndRgba,
+    "add-alpha": fixAddAlpha,
+    transparentize: fixTransparentize,
+    // In SASS we support compilation of functions
+    ...(options.language === "scss"
+      ? {
+          mix: compileSassValueInPlace,
+          darken: compileSassValueInPlace,
+          lighten: compileSassValueInPlace,
+          saturate: compileSassValueInPlace,
+          desaturate: compileSassValueInPlace,
+          "add-tint": compileSassValueInPlace,
+          "add-shade": compileSassValueInPlace,
+          "adjust-hue": compileSassValueInPlace,
+        }
+      : {}),
+  })
+
+  // TODO: Figure out a method that doesn't just bail out when it finds an unmigratable (there could be fixable values within the same declaration)
+  if (containsUnmigratableFunction(newValue)) {
+    options.reporter({
+      autofixAvailable: false,
+      message: unsupportedFunctionMessage,
+      node: decl,
+    })
+  }
+  if (options.fix && newValue !== decl.value && !isVariable(decl)) {
+    decl.replaceWith(
+      decl.clone({
+        value: newValue,
+      })
+    )
+  }
+}
+
+/**
+ * This will lint the functions within a declaration, and migrate anything it can (such as with rgb, rgba and add-alpha)
+ */
+export const noInvalidFunctionsRule = (
+  stylesheetNode: Root,
+  options: Options
+) => {
+  walkDeclsWithKaizenTokens(stylesheetNode, ({ postcssNode }) => {
+    noInvalidFunctionsOnDeclaration(postcssNode, options)
+  })
+}
+export const declContainsInvalidFunctions = (
```

Comment:
Having this export here used in a separate rule makes me think it's in the wrong place and should be in a separate utility module. (Not a deal-breaker if you want to leave it for now in the interest of merging)

Reply:
Agree this is strange, i'm not sure what the best solution is... e.g.
- move just this to a separate file but still import from the no-invalid-functions rule module
- completely refactor the rule so that any "checking" logic is separate from the actual rule, i.e. the rule is a composition of a set of primitives/utilities (i like this the best)

Same goes for no-invalid-equations.

Reply:
The 2nd point might be the way to do. In any case i think this is a job for a subsequent PR.

Reply:
FWIW, I'm happy to merge without this refactor in the interest of getting it done before you go on leave